### PR TITLE
Fix get device information when the application has partial trust mode

### PR DESCRIPTION
**Pull request by AnastasiaKubova**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests if this modifies the Windows code?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

[AB#82519](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/82519)

### CODE REVIEW

File: Tests/Microsoft.AppCenter.Test.WindowsDesktop.NetCore/Utils/DeviceInformationHelperTest.cs [hunk 1]
```diff
@@ -10,13 +12,41 @@ namespace Microsoft.AppCenter.Test.WindowsDesktop.Utils
...
             Assert.Equal("appcenter.wpf.netcore", device.SdkName);
         }
+
+        /// <summary>
+        /// Verify that sdk is not crashed during getting device info when management class is not available.
+        /// </summary>
+        [Fact]
+        public void VerifyDeviceInfoWhenManagmentClassNotAvailable()
+        {
+            DeviceInformationHelper deviceInformation = new DeviceInformationHelper();
```
Comment (MatkovIvan) [inline Tests/Microsoft.AppCenter.Test.WindowsDesktop.NetCore/Utils/DeviceInformationHelperTest.cs:30]:
```suggestion
            var deviceInformation = new DeviceInformationHelper();
```

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs [hunk 2]
```diff
@@ -48,11 +72,19 @@ protected override string GetAppNamespace()
...
         {
-            var managementClass = new ManagementClass("Win32_ComputerSystem");
-            foreach (var managementObject in managementClass.GetInstances())
+            try
+            {
+                ManagementClass managementClass = _managmentClassFactory.GetComputerSystemClass();
+                foreach (var managementObject in managementClass.GetInstances())
+                {
+                    var manufacturer = (string)managementObject["Manufacturer"];
+                    return (string.IsNullOrEmpty(manufacturer) || DefaultSystemManufacturer == manufacturer ? null : manufacturer);
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs:81]:
```suggestion
                    return string.IsNullOrEmpty(manufacturer) || DefaultSystemManufacturer == manufacturer ? null : manufacturer;
```

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs [hunk 3]
```diff
@@ -48,11 +72,19 @@ protected override string GetAppNamespace()
 
         protected override string GetDeviceOemName()
         {
-            var managementClass = new ManagementClass("Win32_ComputerSystem");
-            foreach (var managementObject in managementClass.GetInstances())
+            try
+            {
+                ManagementClass managementClass = _managmentClassFactory.GetComputerSystemClass();
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs:77]:
```suggestion
                var managementClass = _managmentClassFactory.GetComputerSystemClass();
```

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs [hunk 4]
```diff
@@ -32,11 +48,19 @@ protected override string GetSdkName()
...
+                foreach (var managementObject in managementClass.GetInstances())
+                {
+                    var model = (string)managementObject["Model"];
+                    return (string.IsNullOrEmpty(model) || DefaultSystemProductName == model ? null : model);
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs:57]:
```suggestion
                    return string.IsNullOrEmpty(model) || DefaultSystemProductName == model ? null : model;
```

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs [hunk 5]
```diff
@@ -32,11 +48,19 @@ protected override string GetSdkName()
 
         protected override string GetDeviceModel()
         {
-            var managementClass = new ManagementClass("Win32_ComputerSystem");
-            foreach (var managementObject in managementClass.GetInstances())
+            try
+            {
+                ManagementClass managementClass = _managmentClassFactory.GetComputerSystemClass();
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs:53]:
```suggestion
                var managementClass = _managmentClassFactory.GetComputerSystemClass();
```

**Review by russelarms — APPROVED** [review]

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs [hunk 6]
```diff
@@ -90,10 +122,18 @@ protected override string GetOsBuild()
...
+                var managementClass = _managmentClassFactory.GetOperatingSystemClass();
+                foreach (var managementObject in managementClass.GetInstances())
+                {
+                    return (string)managementObject["Version"];
+                }
+            }
+            catch (UnauthorizedAccessException exc)
             {
-                return (string)managementObject["Version"];
+                AppCenterLog.Warn(AppCenterLog.LogTag, "Failed to get device os version with error: ", exc);
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs:135]:
```suggestion
                AppCenterLog.Warn(AppCenterLog.LogTag, "Failed to get device OS version with error: ", exception);
```

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs [hunk 7]
```diff
@@ -90,10 +122,18 @@ protected override string GetOsBuild()
...
-            foreach (var managementObject in managementClass.GetInstances())
+            try
+            {
+                var managementClass = _managmentClassFactory.GetOperatingSystemClass();
+                foreach (var managementObject in managementClass.GetInstances())
+                {
+                    return (string)managementObject["Version"];
+                }
+            }
+            catch (UnauthorizedAccessException exc)
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs:133]:
```suggestion
            catch (UnauthorizedAccessException exception)
```

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs [hunk 8]
```diff
@@ -48,11 +72,19 @@ protected override string GetAppNamespace()
...
+                {
+                    var manufacturer = (string)managementObject["Manufacturer"];
+                    return string.IsNullOrEmpty(manufacturer) || DefaultSystemManufacturer == manufacturer ? null : manufacturer;
+                }
+            } 
+            catch (UnauthorizedAccessException exc)
             {
-                var manufacturer = (string)managementObject["Manufacturer"];
-                return (string.IsNullOrEmpty(manufacturer) || DefaultSystemManufacturer == manufacturer ? null : manufacturer);
+                AppCenterLog.Warn(AppCenterLog.LogTag, "Failed to get device oem name with error: ", exc);
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs:86]:
```suggestion
                AppCenterLog.Warn(AppCenterLog.LogTag, "Failed to get device OEM name with error: ", exception);
```

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs [hunk 9]
```diff
@@ -48,11 +72,19 @@ protected override string GetAppNamespace()
...
+            try
+            {
+                var managementClass = _managmentClassFactory.GetComputerSystemClass();
+                foreach (var managementObject in managementClass.GetInstances())
+                {
+                    var manufacturer = (string)managementObject["Manufacturer"];
+                    return string.IsNullOrEmpty(manufacturer) || DefaultSystemManufacturer == manufacturer ? null : manufacturer;
+                }
+            } 
+            catch (UnauthorizedAccessException exc)
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs:84]:
```suggestion
            catch (UnauthorizedAccessException exception)
```

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/ManagmentClassFactory.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        {
+        }
+
+        public static ManagmentClassFactory GetInstance()
+        {
+            if (Instance == null)
+            {
+                Instance = new ManagmentClassFactory();
+            }
+            return Instance;
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/ManagmentClassFactory.cs:37]:
```suggestion
            return _instance ?? _instance = new ManagmentClassFactory();
```

File: SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/ManagmentClassFactory.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        /// <summary>
+        /// Get object of class Win32_OperatingSystem.
+        /// </summary>
+        /// <returns>Win32_OperatingSystem object.</returns>
+        ManagementClass GetOperatingSystemClass();
+    }
+
+    class ManagmentClassFactory : IManagmentClassFactory
+    {
+        private static ManagmentClassFactory Instance;
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/ManagmentClassFactory.cs:25]:
```suggestion
        private static ManagmentClassFactory _instance;
```

**Review by MatkovIvan — COMMENTED** [review]
A bit more style nitpicks

**Review by MatkovIvan — APPROVED** [review]

File: CHANGELOG.md [hunk 12]
```diff
@@ -7,6 +7,7 @@
 #### WPF/WinForms
 
 * **[Fix]** Fix exception handling when the app tries to access a locked database.
+* **[Fix]** Fix get device information when the application has partial trust mode.
```
Comment (Jamminroot) [inline CHANGELOG.md:10]:
```suggestion
* **[Fix]** Fix getting device information when the application runs in partial trust mode.
```

**Review by Jamminroot — APPROVED** [review]

**Review by annakocheshkova — APPROVED** [review]

**Review by MatkovIvan — APPROVED** [review]

**Pull request merged**